File: core/services.go
// Statping
// Copyright (C) 2018.  Hunter Long and the project contributors
// Written by Hunter Long <info@socialeck.com> and the project contributors
//
// https://github.com/hunterlong/statping
//
// The licenses for most software and other practical works are designed
// to take away your freedom to share and change the works.  By contrast,
// the GNU General Public License is intended to guarantee your freedom to
// share and change all versions of a program--to make sure it remains free
// software for all its users.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package core

import (
	"fmt"
	"github.com/hunterlong/statping/core/notifier"
	"github.com/hunterlong/statping/database"
	"github.com/hunterlong/statping/types"
	"sort"
	"time"
)

type Service struct {
	*database.ServiceObj
}

func Services() map[int64]*Service {
	return CoreApp.services
}

// SelectService returns a *core.Service from in memory
func SelectService(id int64) *Service {
	service := CoreApp.services[id]
	if service != nil {
		return service
	}
	return nil
}

func (s *Service) AfterCreate(obj interface{}, err error) {

}

// CheckinProcess runs the checkin routine for each checkin attached to service
func CheckinProcess(s database.Servicer) {
	for _, c := range s.Checkins() {
		c.Start()
		go CheckinRoutine(c)
	}
}

// SelectAllServices returns a slice of *core.Service to be store on []*core.Services
// should only be called once on startup.
func SelectAllServices(start bool) (map[int64]*Service, error) {
	services := make(map[int64]*Service)
	if len(CoreApp.services) > 0 {
		return CoreApp.services, nil
	}

	for _, s := range database.Services() {
		if start {
			s.Start()
			CheckinProcess(s)
		}

		fails := s.Failures().Last(limitedFailures)
		s.Service.Failures = fails

		for _, c := range s.Checkins() {
			s.Service.Checkins = append(s.Service.Checkins, c.Checkin)
		}

		// collect initial service stats
		s.UpdateStats()

		services[s.Id] = &Service{s}
	}

	CoreApp.services = services
	reorderServices()

	return services, nil
}

func wrapFailures(f []*types.Failure) []*Failure {
	var fails []*Failure
	for _, v := range f {
		fails = append(fails, &Failure{v})
	}
	return fails
}

// reorderServices will sort the services based on 'order_id'
func reorderServices() {
	sort.Sort(ServiceOrder(CoreApp.services))
}

// updateService will update a service in the []*core.Services slice
func updateService(s *Service) {
	CoreApp.services[s.Id] = s
}

// Delete will remove a service from the database, it will also end the service checking go routine
func (s *Service) Delete() error {
	err := database.Delete(s)
	if err != nil {
		log.Errorln(fmt.Sprintf("Failed to delete service %v. %v", s.Name, err))
		return err
	}
	s.Close()
	CoreApp.services[s.Id] = nil
	reorderServices()
	notifier.OnDeletedService(s.Service)
	return err
}

// Update will update a service in the database, the service's checking routine can be restarted by passing true
func Update(s *Service, restart bool) error {
	err := database.Update(s)
	if err != nil {
		log.Errorln(fmt.Sprintf("Failed to update service %v. %v", s.Name, err))
		return err
	}
	// clear the notification queue for a service
	if !s.AllowNotifications.Bool {
		for _, n := range CoreApp.Notifications {
			notif := n.(notifier.Notifier).Select()
			notif.ResetUniqueQueue(fmt.Sprintf("service_%v", s.Id))
		}
	}
	if restart {
		s.Close()
		s.Start()
		s.SleepDuration = s.Duration()
		go ServiceCheckQueue(s, true)
	}
	reorderServices()
	updateService(s)
	notifier.OnUpdatedService(s.Service)
	return err
}

// Create will create a service and insert it into the database
func Create(srv database.Servicer, check bool) (int64, error) {
	s := srv.Model()
	s.CreatedAt = time.Now().UTC()
	_, err := database.Create(s)
	if err != nil {
		log.Errorln(fmt.Sprintf("Failed to create service %v #%v: %v", s.Name, s.Id, err))
		return 0, err
	}
	service := &Service{s}
	s.Start()
	CoreApp.services[service.Id] = service
	go ServiceCheckQueue(service, check)
	reorderServices()
	notifier.OnNewService(s.Service)
	return s.Id, nil
}

File: dev/test/cypress/integration/services.js
/*
 * Statping
 * Copyright (C) 2018.  Hunter Long and the project contributors
 * Written by Hunter Long <info@socialeck.com> and the project contributors
 *
 * https://github.com/hunterlong/statping
 *
 * The licenses for most software and other practical works are designed
 * to take away your freedom to share and change the works.  By contrast,
 * the GNU General Public License is intended to guarantee your freedom to
 * share and change all versions of a program--to make sure it remains free
 * software for all its users.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

context('Service Tests', () => {

    beforeEach(function () {
        cy.visit('http://localhost:8080/dashboard')
        cy.get('input[name="username"]').type('admin')
        cy.get('input[name="password"]').type('admin')
        cy.get('form').submit()
    })

    it('should view services', () => {
        cy.visit('http://localhost:8080/services')
        cy.get('tr').should('have.length', 6)
        cy.title().should('eq', 'Statping | Services')
    })

    it('should create HTTP GET service', () => {
        cy.visit('http://localhost:8080/services')
        cy.get('select[name="method"]').select('GET')
        cy.get('input[name="name"]').clear().type('Google.com')
        cy.get('select[name="check_type"]').select('http')
        cy.get('input[name="domain"]').clear().type('https://google.com')
        cy.get('input[name="expected_status"]').clear().type('200')
        cy.get('input[name="interval"]').clear().type('25')
        cy.get('input[name="timeout"]').clear().type('30')
        cy.get('form').submit()
        cy.title().should('eq', 'Statping | Services')
        cy.get('tr').should('have.length', 7)
    })

    it('should create HTTP POST service', () => {
        cy.visit('http://localhost:8080/services')
        cy.get('select[name="method"]').select('POST')
        cy.get('input[name="name"]').clear().type('JSON Regex Test')
        cy.get('select[name="check_type"]').select('http')
        cy.get('input[name="domain"]').clear().type('https://jsonplaceholder.typicode.com/posts')
        cy.get('textarea[name="post_data"]').clear().type(`(title)": "((\\"|[statping])*)"`)
        cy.get('input[name="expected_status"]').clear().type('201')
        cy.get('input[name="interval"]').clear().type('15')
        cy.get('input[name="timeout"]').clear().type('45')
        cy.get('form').submit()
        cy.title().should('eq', 'Statping | Services')
        cy.get('tr').should('have.length', 8)
    })

    it('should create TCP service', () => {
        cy.visit('http://localhost:8080/services')
        cy.get('select[name="check_type"]').select('tcp')
        cy.get('input[name="name"]').clear().type('Google DNS')
        cy.get('input[name="domain"]').clear().type('8.8.8.8')
        cy.get('input[name="port"]').clear().type('53')
        cy.get('input[name="interval"]').clear().type('25')
        cy.get('input[name="timeout"]').clear().type('15')
        cy.get('form').submit()
        cy.title().should('eq', 'Statping | Services')
        cy.get('tr').should('have.length', 9)
    })

    it('should view HTTP GET service', () => {
        cy.visit('http://localhost:8080/service/6')
        cy.title().should('eq', 'Statping | Google.com Service')
    })

    it('should view HTTP POST service', () => {
        cy.visit('http://localhost:8080/service/7')
        cy.title().should('eq', 'Statping | JSON Regex Test Service')
    })

    it('should view TCP service', () => {
        cy.visit('http://localhost:8080/service/8')
        cy.title().should('eq', 'Statping | Google DNS Service')
    })

    it('should update HTTP service', () => {
        cy.visit('http://localhost:8080/service/6')
        cy.title().should('eq', 'Statping | Google.com Service')
        cy.get('#service_name').clear().type('Google Updated')
        cy.get('#service_interval').clear().type('60')
        cy.get(':nth-child(3) > form').submit()
        cy.title().should('eq', 'Statping | Google Updated Service')
        cy.get('#service_name').should('have.value', 'Google Updated')
    });

    it('should check the updated service', () => {
        cy.visit('http://localhost:8080/service/6')
        cy.title().should('eq', 'Statping | Google Updated Service')
        cy.get('#service_name').should('have.value', 'Google Updated')
        cy.get('#service_interval').should('have.value', '60')
    })

    it('should delete a service', () => {
        cy.visit('http://localhost:8080/services')
        cy.get(':nth-child(5) > .text-right > .btn-group > .btn-danger').click()
        cy.title().should('eq', 'Statping | Services')
        cy.get('tr').should('have.length', 8)
    })


});
